media_gateway_server/src/server/service/cache.rs

```diff
@@ -0,0 +1,138 @@
+use std::sync::{Arc, OnceLock};
+use std::time::{Duration, Instant};
+
+use anyhow::bail;
+use mockall::automock;
+use parking_lot::Mutex;
+use tokio::sync::oneshot::Receiver;
+use tokio::time::sleep;
+
+pub struct CacheStatistics {
+    evicted: u32,
+}
+
+impl CacheStatistics {
+    pub fn new() -> Self {
+        CacheStatistics { evicted: 0 }
+    }
+
+    pub fn register_evicted(&mut self) {
+        match self.evicted.checked_add(1) {
+            Some(result) => self.evicted = result,
+            None => log::error!("Evicted statistics overflow"),
+        }
+    }
+
+    pub fn reset(&mut self) -> Self {
+        let evicted = self.evicted;
+        self.evicted = 0;
+        CacheStatistics { evicted }
+    }
+}
+
+#[automock]
+pub trait CacheUsageTracker {
+    fn register_evicted(&self);
+
+    fn reset(&self) -> CacheStatistics;
+}
+
+pub struct NopeCacheUsageTracker {}
+
+impl CacheUsageTracker for NopeCacheUsageTracker {
+    fn register_evicted(&self) {
+        // do nothing
+    }
+
+    fn reset(&self) -> CacheStatistics {
+        CacheStatistics::new()
+    }
+}
+
+pub struct CacheUsageTrackerImpl {
+    statistics: Arc<Mutex<CacheStatistics>>,
+}
+
+impl CacheUsageTrackerImpl {
+    pub fn new() -> Self {
+        CacheUsageTrackerImpl {
+            statistics: Arc::new(Mutex::new(CacheStatistics::new())),
+        }
+    }
+}
+
+impl CacheUsageTracker for CacheUsageTrackerImpl {
+    fn register_evicted(&self) {
+        let mut statistics = self.statistics.lock();
+        statistics.register_evicted()
+    }
+
+    fn reset(&self) -> CacheStatistics {
+        let mut statistics = self.statistics.lock();
+        statistics.reset()
+    }
+}
+
+pub trait CacheUsageService {
+    async fn run(&self, shutdown_signal: Receiver<()>) -> anyhow::Result<()>;
+}
+
+pub struct EvictedThresholdWarningCacheUsageService {
+    period: Duration,
+    threshold: f64,
+    cache_usage_tracker: Arc<Box<dyn CacheUsageTracker + Send + Sync>>,
```

Comment:
dyn is required because CacheUsageTracker is a trait. The same approach is used for Cache. Seems reasonable to use the same approach for both Cache and EvictedThreshholdWarningCacheUsageService. Generics is not a good idea for Cache as we need to know the exact type in order to use it in security

> |         req.app_data::<Data<Cache<Credentials, BasicAuthCheckResult, _>>>();
   |             ^^^^^^^^ cannot infer type of the type parameter `T` declared on the method `app_data`

or
> error[E0277]: the size for values of type `dyn CacheUsageTracker` cannot be known at compilation time
  --> media_gateway_server/src/server/security.rs:80:24
   |
80 |         req.app_data::<Data<Cache<Credentials, BasicAuthCheckResult, dyn CacheUsageTracker>>>();
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `dyn CacheUsageTracker`
note: required by an implicit `Sized` bound in `Cache`
  --> media_gateway_server/src/server/service/cache.rs:16:24
   |
16 | pub struct Cache<K, V, S> {
   |                        ^ required by the implicit `Sized` requirement on this type parameter in `Cache`
help: consider relaxing the implicit `Sized` restriction
  --> media_gateway_server/src/server/service/cache.rs:16:25
   |
16 | pub struct Cache<K, V, S: ?Sized> {
   |                         ++++++++